libraries/eosiolib/contracts/eosio/key_value.hpp

```diff
@@ -0,0 +1,863 @@
+         auto copy_size = internal_use_do_not_use::kv_get_data(db, 0, (char*)buffer, value_size);
+
+         datastream<const char*> ds((char*)buffer, value_size);
+
+         T val;
+         ds >> val;
+         if (value_size > detail::max_stack_buffer_size) {
+            free(buffer);
+         }
+         return val;
+      }
+
+      /**
+       * Returns an iterator pointing to the element with the lowest key greater than or equal to the given key.
+       * @ingroup keyvalue
+       *
+       * @return An iterator pointing to the element with the lowest key greater than or equal to the given key.
+       */
+      template <typename K>
+      iterator lower_bound(K key) {
+         auto prefix = make_prefix(table_name, name);
+         auto t_key = table_key(prefix, make_key(key));
+
+         uint32_t itr = internal_use_do_not_use::kv_it_create(db, contract_name.value, prefix.data(), prefix.size());
+         int32_t itr_stat = internal_use_do_not_use::kv_it_lower_bound(itr, t_key.data(), t_key.size());
+
+         if (static_cast<kv_it_stat>(itr_stat) == kv_it_stat::iterator_end) {
+            return end();
+         }
+
+         return {contract_name, itr, static_cast<kv_it_stat>(itr_stat), this};
+      }
+
+      /**
+       * Returns an iterator pointing to the element with the highest key less than or equal to the given key.
+       * @ingroup keyvalue
+       *
+       * @return An iterator pointing to the element with the highest key less than or equal to the given key.
+       */
+      template <typename K>
+      iterator upper_bound(K key) {
+         auto prefix = make_prefix(table_name, name);
+         auto t_key = table_key(prefix, make_key(key));
+
+         uint32_t itr = internal_use_do_not_use::kv_it_create(db, contract_name.value, prefix.data(), prefix.size());
+         int32_t itr_stat = internal_use_do_not_use::kv_it_lower_bound(itr, t_key.data(), t_key.size());
+
+         iterator it{contract_name, itr, static_cast<kv_it_stat>(itr_stat), this};
+
+         auto cmp = internal_use_do_not_use::kv_it_key_compare(it.itr, t_key.data(), t_key.size());
+         while(cmp > 0) {
+            if (it == begin()) {
+               return end();
+            }
+            --it;
+            cmp = internal_use_do_not_use::kv_it_key_compare(it.itr, t_key.data(), t_key.size());
+         }
+
+         return it;
+
+      }
+
+      /**
+       * Returns an iterator referring to the `past-the-end` element. It does not point to any element, therefore `value` should not be called on it.
+       * @ingroup keyvalue
+       *
+       * @return An iterator referring to the `past-the-end` element.
+       */
+      iterator end() {
+         auto prefix = make_prefix(table_name, name);
+         uint32_t itr = internal_use_do_not_use::kv_it_create(db, contract_name.value, prefix.data(), prefix.size());
+         int32_t itr_stat = internal_use_do_not_use::kv_it_move_to_end(itr);
+
```

Comment:
Perfomance issues:
* `kv_it_create` creates an iterator at the end. `kv_it_move_to_end` is redundant.
* The following pattern creates and destroys intrinsic iterators each iteration through the loop:

```
for(auto it = x.begin(); it != x.end(); ++it)
```

A potential fix: `end()` could return a `kv_index::iterator` with a null intrinsic iterator.

Reply:
Is decrementing an end iterator considered valid? Or is an end iterator only ever allowed to point off the end?

Reply:
C++: if begin != end, then decrementing end is valid

Reply:
kv intrinsics: incrementing and decrementing end is always valid. If nothing matches the prefix, then the result is end (circles back on itself).

Reply:
Okay so if we had end store a null iterator, we would need to create an iterator inside the `--` operator if one didn't already exist.

Reply:
yes